seastar/src/distributed.rs

```diff
@@ -0,0 +1,537 @@
+    ///
+    /// impl Service for CounterService {
+    ///     fn stop(&self) -> Box<dyn Future<Output = ()>> {
+    ///         let counter = self.0.clone();
+    ///         Box::new(async move {
+    ///             counter.fetch_add(1, Ordering::SeqCst);
+    ///         })
+    ///     }
+    /// }
+    ///
+    /// #[seastar::test]
+    /// async fn test_start_and_stop() {
+    ///     let counter: Arc<AtomicU32> = Default::default();
+    ///     let counter_clone = counter.clone();
+    ///     let service_maker = move || CounterService(counter_clone.clone());
+    ///     let distr = Distributed::start(service_maker).await;
+    ///     distr.stop().await;
+    ///     assert_eq!(get_count(), counter.load(Ordering::SeqCst));
+    /// }
+    /// ```
+    pub fn start<Func>(service_maker: Func) -> impl Future<Output = Self>
+    where
+        Func: Fn() -> S + Sync,
+    {
+        Distributed::start_inner(service_maker, false)
+    }
+
+    /// Stops the service on all shards on which it was ran, freeing each instance's memory. Effectively an async destructor.
+    ///
+    /// This **must** be called when the distributed service is no longer to be used!.
+    ///
+    /// ```rust
+    /// use std::future::Future;
+    /// use std::sync::atomic::{AtomicU32, Ordering};
+    /// use std::sync::Arc;
+    /// use seastar::{Distributed, Service};
+    ///
+    /// struct CounterService(Arc<AtomicU32>);
+    ///
+    /// impl Service for CounterService {
+    ///     fn stop(&self) -> Box<dyn Future<Output = ()>> {
+    ///         let counter = self.0.clone();
+    ///         Box::new(async move {
+    ///             counter.fetch_add(1, Ordering::SeqCst);
+    ///         })
+    ///     }
+    /// }
+    ///
+    /// #[seastar::test]
+    /// async fn test_start_single_and_stop() {
+    ///     let counter_clone = counter.clone();
+    ///     let service_maker = move || CounterService(counter_clone.clone());
+    ///     let distr = Distributed::start_single(service_maker).await;
+    ///     distr.stop().await;
+    ///     assert_eq!(1, counter.load(Ordering::SeqCst));
+    /// }
+    /// ```
+    pub async fn stop(&self) {
+        // TODO: is calling this manually a chore? If so, we could implement `Drop` as follows:
+        //```
+        // impl<S: Service> Drop for Distributed<S> {
+        //     fn drop(&mut self) {
+        //         // You can't use `async` in a `Drop` implementation,
+        //         // so you need to block the current thread to wait for the `stop` method to complete.
+        //         futures::executor::block_on(async {
+        //             self.stop().await;
+        //         });
+        //     }
+        // }
+        //```
+        // If we decide to do so, it'd make sense to change the signature of `stop` to `async fn stop(&mut self)`.
+        // This however would produce an inconsistency, as the `start*` methods operate on interior mutability (take by `&self`).
```

Comment:
No, we musn't block in Drop. Doing this would block the shard's reactor and would prevent other tasks from running - which would cause a deadlock if `stop()` spawns some tasks and then waits for them.